machine-check/README.md

```diff
@@ -1,3 +1,123 @@
+- the name of the machine role that is allowed to issue this command, e.g. `Control`
+- a list of event types that record this transition, e.g. `Opening`
+- the state we thus arrive at, e.g. `(Opening)`
+
+The short form for writing this down is `(Closed) --[open@Control<Opening>]--> (Opening)`
+
+## Example protocol
+
+The machines from the [Hangar Door example](../machine-runner/README.md#example-usage) might follow this protocol:
+
+- `(Closed) --[open@Control<Opening>]--> (Opening)`
+- `(Opening) --[update@Door<Opening>]--> (Opening)`
+- `(Opening) --[open@Door<Opened>]--> (Open)`
+- `(Open) --[close@Control<Closing>]--> (Closing)`
+- `(Closing) --[update@Door<Closing>]--> (Closing)`
+- `(Closing) --[close@Door<Closed>]--> (Closed)`
+
+The Control can initiate opening and closing while the Door provides progress updates and states when each movement has been completed.
+
+## How to use this library
+
+This library is typically used within your unit tests to check the structure of the machines you’ve written.
+For this we need to provide two pieces: the desired swarm protocol and the event subscriptions of your machine roles.
+Continuing the example above, it could look like this:
+
+```ts
+import { Door, Control, HangarBay } from './example-proto.js'
+import { SwarmProtocolType, checkProjection, checkSwarmProtocol } from '@actyx/machine-check'
+
+const swarmProtocol: SwarmProtocolType = {
+  initial: 'Closed',
+  transitions: [
+    {
+      source: 'Closed',
+      target: 'Opening',
+      label: { cmd: 'open', role: 'Control', logType: ['opening'] },
+    },
+    {
+      source: 'Opening',
+      target: 'Opening',
+      label: { cmd: 'update', role: 'Door', logType: ['opening'] },
+    },
+    {
+      source: 'Opening',
+      target: 'Open',
+      label: { cmd: 'open', role: 'Door', logType: ['opened'] },
+    },
+    {
+      source: 'Open',
+      target: 'Closing',
+      label: { cmd: 'close', role: 'Control', logType: ['closing'] },
+    },
+    {
+      source: 'Closing',
+      target: 'Closing',
+      label: { cmd: 'update', role: 'Door', logType: ['closing'] },
+    },
+    {
+      source: 'Closing',
+      target: 'Closed',
+      label: { cmd: 'close', role: 'Door', logType: ['closed'] },
+    },
+  ],
+}
+
+const subscriptions = {
+  Control: ['closing', 'closed', 'opening', 'opened'],
+  Door: ['closing', 'closed', 'opening', 'opened'],
+}
+
+console.log(checkSwarmProtocol(swarmProtocol, subscriptions))
+console.log(
+  checkProjection(
+    swarmProtocol,
+    subscriptions,
+    'Control',
+    Control.Control.createJSONForAnalysis(Control.Closed),
+  ),
+)
+console.log(
+  checkProjection(
+    swarmProtocol,
+    subscriptions,
+    'Door',
+    Door.Door.createJSONForAnalysis(Door.Closed),
+  ),
+)
+```
+
+Instead of `console.log()` you’d normally assert that the result is `{"type":"OK"}`.
+In the example as given you’ll instead be notified that the overall protocol has a flaw:
```

Comment:
I'd also add an explicit "this is how you run this code" as example.

I don't know about SW's team constitution but I'd argue that adding what might seem "really small details" cut the time to get new people on board.

Further in time we can add examples with Jest or similar testing frameworks but just getting "you can copy this code and these instructions and you've checked your first protocol" is very useful.

Reply:
I think it was from the example of line 48. I agree that this part require more explicit story on why "reuse" is there in the first place.